src/uct/api/uct.h

```diff
@@ -2057,6 +2058,125 @@ ucs_status_t uct_ep_connect_to_ep(uct_ep_h ep, const uct_device_addr_t *dev_addr
+    UCT_MEM_ALLOC_PARAM_FIELD_ADDR_PTR       = UCS_BIT(1),
+
+    /** Enables @ref uct_mem_alloc_param_t::length_p */
+    UCT_MEM_ALLOC_PARAM_FIELD_LENGTH_PTR     = UCS_BIT(2),
+
+    /** Enables @ref uct_mem_alloc_param_t::methods */
+    UCT_MEM_ALLOC_PARAM_FIELD_METHODS        = UCS_BIT(3),
+
+    /** Enables @ref uct_mem_alloc_param_t::num_methods */
+    UCT_MEM_ALLOC_PARAM_FIELD_NUM_METHODS    = UCS_BIT(4),
+
+    /** Enables @ref uct_mem_alloc_param_t::mds */
+    UCT_MEM_ALLOC_PARAM_FIELD_MDS            = UCS_BIT(5),
+
+    /** Enables @ref uct_mem_alloc_param_t::num_mds */
+    UCT_MEM_ALLOC_PARAM_FIELD_NUM_MDS        = UCS_BIT(6),
+
+    /** Enables @ref uct_mem_alloc_param_t::mem_type */
+    UCT_MEM_ALLOC_PARAM_FIELD_MEM_TYPE       = UCS_BIT(7),
+
+    /** Enables @ref uct_mem_alloc_param_t::name */
+    UCT_MEM_ALLOC_PARAM_FIELD_NAME           = UCS_BIT(8)
+} uct_mem_alloc_param_field_t;
+
+
+/**
+ * @ingroup UCT_MD
+ * @brief Parameters for allocating memory using @ref uct_mem_alloc,
+ *        @ref uct_md_mem_alloc
+ */
+typedef struct {
+    /**
+     * Mask of valid fields in this structure, using bits from
+     * @ref uct_mem_alloc_param_field_t. Fields not specified in this mask will
+     * be ignored
+     */
+    uint64_t           field_mask;
+
+    /**
+     * Memory allocation flags, see @ref uct_md_mem_flags
+     */
+    unsigned           flags;
+
+    /**
+     * The address. Valid to use with @ref uct_md_mem_alloc
+     */
+
+    /**
+     * Points to minimal size to allocate. The actual size may be larger, for
+     * example because of alignment restrictions. Upon successful return,
+     * filled with the actual size that was allocated. Must be >0.
+     */
+    size_t             *length_p;
+
+    /**
+     * If @a address_p content is NULL, the underlying allocation routine will
+     * choose the address at which to create the mapping. If @a address_p
+     * content is non-NULL but UCT_MD_MEM_FLAG_FIXED is not set, the address
+     * will be interpreted as a hint as to where to establish the mapping. If
+     * @a address_p is non-NULL and UCT_MD_MEM_FLAG_FIXED is set, then the
+     * specified address is interpreted as a requirement. In this case, if the
+     * mapping to the exact address cannot be made, the allocation request
+     * fails.
+     */
+    void               **address_p;
+
+    /**
+     * Array of memory allocation methods to attempt. Every one of the provided
+     * allocation methods will be used, in turn, to perform the allocation,
+     * until one succeeds. Whenever the MD method is encountered, every one
+     * of the provided MDs will be used, in turn, to allocate the memory, until
+     * one succeeds, or they are exhausted. In this case the next allocation
+     * method from the initial list will be attempted.
+     */
+    uct_alloc_method_t *methods;
+
+    /**
+     * Length of 'methods' array
+     */
+    unsigned           num_methods;
```

Comment:
should put methods and num methods in sub-struct, with one bit in field_mask for both
```
struct {
    const uct_alloc_method_t   *methods;
    unsigned                   count;
} methods;
```

same thing for "mds"